File: LeetCodeRecord/ACFirst/96-NumTrees.java
package ACFirst;

class NumTrees {
    public int numTrees(int n) {
        if (n <= 1) {
            return 1;
        }

        // dp 数组
        int[] dp = new int[n + 1];

        // base case
        dp[0] = 1;
        dp[1] = 1;

        for (int i = 2; i <= n; i++) {
            for (int j = 1; j <= i; j++) {
                dp[i] += dp[j - 1] * dp[i - j];
            }    
        }
        
        return dp[n];
    }

    public static void main(String[] args) {
        NumTrees nt = new NumTrees();
        System.out.println(nt.numTrees(3));
    }
}

File: Top200/96-NumTrees.go
package main

func numTrees(n int) int {
	dp := make([]int, n+1)

	dp[0] = 1
	dp[1] = 1

	for i := 2; i < n+1; i++ {
		for j := 0; j < i; j++ {
			dp[i] += dp[j] * dp[i-j-1]
		}
	}

	return dp[n]
}
